File: app/src/main/java/com/waitou/towards/model/MainActivity.java
package com.waitou.towards.model;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.NavigationView;
import android.support.v4.view.GravityCompat;
import android.view.LayoutInflater;
import android.view.MenuItem;

import com.jaeger.library.StatusBarUtil;
import com.waitou.lib_theme.ChangeModeController;
import com.waitou.lib_theme.ThemeUtils;
import com.waitou.towards.R;
import com.waitou.towards.databinding.ActivityMainBinding;
import com.waitou.towards.databinding.NavHeaderMainBinding;
import com.waitou.towards.model.jokes.activity.RecommendedActivity;
import com.waitou.towards.model.jokes.contract.MainContract;
import com.waitou.towards.model.jokes.fragment.CircleFragment;
import com.waitou.towards.model.jokes.fragment.FigureFragment;
import com.waitou.towards.model.jokes.fragment.HomeFragment;
import com.waitou.towards.model.jokes.fragment.PersonFragment;
import com.waitou.towards.model.jokes.fragment.joke.TextJokeFragment;
import com.waitou.towards.model.presenter.MainPresenter;

import cn.droidlover.xdroid.base.XActivity;
import cn.droidlover.xdroid.base.XFragmentAdapter;


public class MainActivity extends XActivity<MainPresenter, ActivityMainBinding> implements NavigationView.OnNavigationItemSelectedListener, MainContract.MainView {
    private HomeFragment     mHomeFragment;
    private TextJokeFragment mTextJokeFragment;
    private FigureFragment   mFigureFragment;
    private CircleFragment   mCircleFragment;
    private PersonFragment   mPersonFragment;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        ChangeModeController.get().init(this);
        super.onCreate(savedInstanceState);
    }

    @Override
    protected MainPresenter createPresenter() {
        return new MainPresenter(getTextJokeFragment());
    }

    @Override
    public boolean initXView() {
        return false;
    }

    @Override
    public int getContentViewId() {
        return R.layout.activity_main;
    }

    @Override
    public void initData(Bundle savedInstanceState) {
        getBinding().include.toolbar.initMenuActionBar("首页");
        getBinding().include.toolbar.setBackListener(R.drawable.icon_menu, v -> getBinding().mainDrawerLayout.openDrawer(GravityCompat.START));
        XFragmentAdapter adapter = new XFragmentAdapter(getSupportFragmentManager(), getHomeFragment(), getTextJokeFragment(), getFigureFragment(), getCircleFragment(), getPersonFragment());
        getBinding().setAdapter(adapter);
        getBinding().mainTab.setupWithViewPager(getBinding().fContent);
        getBinding().mainTab.enableShiftingMode(false);
        getBinding().fContent.setOffscreenPageLimit(4);
        NavHeaderMainBinding binding = NavHeaderMainBinding.inflate((LayoutInflater) getSystemService(LAYOUT_INFLATER_SERVICE));
        binding.setDrawableId(R.drawable.nav_header_img);
        getBinding().navView.addHeaderView(binding.getRoot());
        getBinding().navView.setNavigationItemSelectedListener(this);
        StatusBarUtil.setColorNoTranslucentForDrawerLayout(MainActivity.this, getBinding().mainDrawerLayout, ThemeUtils.getThemeAttrColor(this, R.attr.colorPrimary));
        ChangeModeController.get().setOnThemeChangeListener(resourceId -> {
            //主题改版之后，刷新一下bottomBar的状态
            getBinding().mainTab.setItemIconTintList(ThemeUtils.getColorStateList(this, R.color.skin_bottom_bar_not));
            getBinding().mainTab.setItemTextColor(ThemeUtils.getColorStateList(this, R.color.skin_bottom_bar_not));
            getBinding().mainTab.setItemBackgroundResource(ThemeUtils.getAttrTypedValue(this, R.attr.colorPrimary).resourceId);
            StatusBarUtil.setColorNoTranslucentForDrawerLayout(MainActivity.this, getBinding().mainDrawerLayout, ThemeUtils.getThemeAttrColor(this, R.attr.colorPrimary));
        });
        getBinding().mainTab.setOnNavigationItemSelectedListener(item -> {
            switch (item.getItemId()) {
                case R.id.menu_home:
                    getBinding().include.toolbar.addCustomMenuView(null, 0);
                    break;
                case R.id.menu_joke:
                    getBinding().include.toolbar.addCustomMenuView(getTextJokeFragment().getJokeToolbar(), R.id.menu_joke);
                    break;
            }
            return true;
        });

    }

    @Override
    public void reloadData() {
    }

    @Override
    public void setPresenter(MainPresenter presenter) {
    }

    /**
     * 左侧菜单是否被打开
     */
    private boolean toggle() {
        return getBinding().mainDrawerLayout.isDrawerOpen(GravityCompat.START);
    }

    @Override
    public void onBackPressed() {
        if (toggle()) {
            getBinding().mainDrawerLayout.closeDrawer(GravityCompat.START);
        } else {
            super.onBackPressed();
        }
    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.nav_tuijian:
                toActivity(RecommendedActivity.class);
                break;
            case R.id.nav_all:
                toActivity(ThemeActivity.class);
                break;
            case R.id.nav_meizi:
                break;
            case R.id.nav_collect:
                break;
            case R.id.nav_theme:
                break;
            case R.id.nav_about:
                break;
        }
        getBinding().mainDrawerLayout.closeDrawer(GravityCompat.START);
        return true;
    }

    /*--------------- 初始化fragment ---------------*/
    private HomeFragment getHomeFragment() {
        if (mHomeFragment == null) {
            mHomeFragment = new HomeFragment();
        }
        return mHomeFragment;
    }

    private TextJokeFragment getTextJokeFragment() {
        if (mTextJokeFragment == null) {
            mTextJokeFragment = new TextJokeFragment();
        }
        return mTextJokeFragment;
    }

    private FigureFragment getFigureFragment() {
        if (mFigureFragment == null) {
            mFigureFragment = new FigureFragment();
        }
        return mFigureFragment;
    }

    private CircleFragment getCircleFragment() {
        if (mCircleFragment == null) {
            mCircleFragment = new CircleFragment();
        }
        return mCircleFragment;
    }

    private PersonFragment getPersonFragment() {
        if (mPersonFragment == null) {
            mPersonFragment = new PersonFragment();
        }
        return mPersonFragment;
    }
}

File: app/src/main/java/com/waitou/towards/model/main/MainActivity.kt
package com.waitou.towards.model.main

import android.app.ActivityManager
import android.content.Intent
import android.os.Bundle
import android.view.Gravity
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import androidx.appcompat.app.AppCompatDelegate
import androidx.appcompat.app.SkinAppCompatDelegateImpl
import androidx.core.content.ContextCompat
import androidx.core.view.GravityCompat
import com.google.android.material.navigation.NavigationView
import com.to.aboomy.statusbar_lib.StatusBarUtil
import com.to.aboomy.theme_lib.annotation.Skinable
import com.to.aboomy.theme_lib.compat.SkinCompatSupportable
import com.to.aboomy.theme_lib.utils.ThemeUtils
import com.umeng.socialize.UMShareAPI
import com.waitou.basic_lib.adapter.BasePagerFragmentAdapter
import com.waitou.imgloader_lib.ImageLoader
import com.waitou.normal_dialog_lib.SheetAdapterDialog
import com.waitou.towards.R
import com.waitou.towards.model.activity.ColorActivity
import com.waitou.towards.model.activity.FlutterActivity
import com.waitou.towards.model.activity.GloadActivity
import com.waitou.towards.model.gallery.GalleryNewActivity
import com.waitou.towards.model.graffiti.GraffitiActivity
import com.waitou.towards.model.main.adater.StylesAdapter
import com.waitou.towards.model.main.fragment.home.HomeNewFragment
import com.waitou.towards.model.main.fragment.joke.JokeFragment
import com.waitou.wt_library.base.BaseActivity
import com.waitou.wt_library.router.Router
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import kotlinx.android.extensions.CacheImplementation
import kotlinx.android.extensions.ContainerOptions
import kotlinx.android.synthetic.main.activity_main.*
import java.util.concurrent.TimeUnit

@Skinable
@ContainerOptions(CacheImplementation.SPARSE_ARRAY)
class MainActivity : BaseActivity(), NavigationView.OnNavigationItemSelectedListener, SkinCompatSupportable {

    private val homeFragment by lazy { HomeNewFragment() }
    private val jokeFragment by lazy { JokeFragment() }

    override fun immersiveStatusBar(): Boolean {
        return false
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        pageTitleBar.setPadding(0, StatusBarUtil.getStatusBarHeight(this), 0, 0)
        StatusBarUtil.drawerLayoutForColor(this, ContextCompat.getColor(this, R.color.bg_grey), drawerLayout)
        val adapter = BasePagerFragmentAdapter(supportFragmentManager, homeFragment, jokeFragment)
        fContent.offscreenPageLimit = 4
        fContent.adapter = adapter
        mainTab.setupWithViewPager(fContent)
        val navHeaderView = ff<ImageView>(R.layout.nav_header_main)
        ImageLoader.displayImage(navHeaderView, R.drawable.nav_header_img)
        navView.addHeaderView(navHeaderView)
        navView.setNavigationItemSelectedListener(this)
        mainTab.setOnNavigationItemSelectedListener {
            when (it.itemId) {
                R.id.menu_home -> pageTitleBar.visibility = View.GONE
                R.id.menu_joke -> pageTitleBar.replaceTitleView(jokeFragment.jokeTitleBar)
//                R.id.menu_figure -> pageTitleBar.restoreTitleView("趣图")
//                R.id.menu_circle -> pageTitleBar.restoreTitleView("圈子")
//                R.id.menu_personal -> pageTitleBar.restoreTitleView("我的")
            }
            true
        }
    }

    override fun getDelegate(): AppCompatDelegate {
        return SkinAppCompatDelegateImpl.get(this, this)
    }

    override fun onBackPressed() {
        if (drawerLayout.isDrawerOpen(GravityCompat.START)) {
            drawerLayout.closeDrawer(GravityCompat.START)
        } else {
            super.onBackPressed()
        }
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        drawerLayout.closeDrawer(GravityCompat.START)
        pend(Observable.timer(200, TimeUnit.MILLISECONDS, AndroidSchedulers.mainThread())
                .subscribe {
                    when (item.itemId) {
                        R.id.nav_tuijian -> Router.newIntent().from(this).to(ColorActivity::class.java).launch()
                        R.id.nav_all -> {
                            Router.newIntent().from(this).to(GloadActivity::class.java).launch()

                        }
                        R.id.nav_meizi -> Router.newIntent().from(this).to(GalleryNewActivity::class.java).launch()
                        R.id.nav_graffiti -> Router.newIntent().from(this).to(GraffitiActivity::class.java).launch()
//                        R.mediaId.nav_collect -> PhotoPickerFinal.getInstance()
//                                .with(this)
//                                .setStrPhotoList("http://img.hb.aicdn.com/621034b37c53ffc81f5d6a23ae1226d5c67e2b9628267-BYuZLo_fw658", "http://img.hb.aicdn.com/e8cbd6ac1b44cf290debbf1ebcdfac6bdf20487f46146-uz70dU_fw658", "http://img.hb.aicdn.com/0d6adb99906f5dc5107962b8446623ea17b1be4d37679-oKTIYt_fw658", "http://img.hb.aicdn.com/d63edb11718f59390c92b17fe9399215c4f7c96c28643-CXnznk_fw658")
//                                .executePreViewPhoto()
                        R.id.nav_theme -> changeNight()
                        R.id.nav_about -> startActivity(Intent(this, FlutterActivity::class.java))
                    }
                })
        return true
    }

    override fun applySkin() {
        StatusBarUtil.drawerLayoutForColor(this, ContextCompat.getColor(this, R.color.bg_grey), drawerLayout)
        mainTab.itemIconTintList = ThemeUtils.getColorStateList(this, R.color.skin_bottom_bar_not)
        mainTab.itemTextColor = ThemeUtils.getColorStateList(this, R.color.skin_bottom_bar_not)
        homeFragment.applySkin()
    }

    /**
     * 主题更换
     */
    private fun changeNight() {
        drawerLayout.closeDrawer(GravityCompat.START)
        val adapter = StylesAdapter(this)
        val dialog = SheetAdapterDialog()
                .grid(3)
                .setTitle("更换主题")
                .setRecyclerAdapter(adapter)
                .setGravity(Gravity.CENTER)
        dialog.show(supportFragmentManager)
        adapter.call = {
            dialog.dismiss()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        UMShareAPI.get(this).onActivityResult(requestCode, resultCode, data)
    }

    override fun onDestroy() {
        super.onDestroy()
        UMShareAPI.get(this).release()
        killAll()
    }

    /**
     * 杀死进程
     */
    private fun killAll() {
        android.os.Process.killProcess(android.os.Process.myPid())
        val manager = getSystemService(ACTIVITY_SERVICE) as ActivityManager
        manager.killBackgroundProcesses(packageName)
    }
}
